contrib/plugs/plug_dependencies.py

```diff
@@ -0,0 +1,137 @@
+"""Extension to OpenHTF to provide dependency declaration for Plugs.
+
+To declare a dependency, add an attribute to the dependent class with the
+PlugDependency descriptor:
+  class NeededPlug(plugs.BasePlug):
+    pass
+
+  class AnotherPlug(plugs.BasePlug):
+    needed = PlugDependency(NeededPlug)
+
+The depended upon plug is then available after the initialization takes place,
+which is done by the phase created by create_dependency_init_phase.
+
+The dependent plugs must not expect their dependencies to work during their
+__init__ calls.
+
+This makes no guarantee about when plugs are torn down; therefore, dependent
+plugs must not use their dependencies during their own tear down.
+"""
+
+import inspect
+
+import openhtf as htf
+from openhtf import plugs
+
+
+class PlugDependency(object):
+  """Descriptor that declares a dependency to another plug."""
+
+  def __init__(self, plug_type):
+    assert issubclass(plug_type, plugs.BasePlug), (
+        '%s must be a subclass of plugs.BasePlug' % plug_type)
+    self.plug_type = plug_type
+    self._attr = '__dependent_plug.%s' % self.plug_type.__name__
+
+  def __get__(self, instance, owner):
+    # If called on the class, return myself for enumeration.
+    if instance is None:
+      return self
+    return getattr(instance, self._attr, None)
+
+  def __set__(self, instance, value):
+    setattr(instance, self._attr, value)
+
+
+def _enumerate_dependencies(plug_type):
+  for attr_name, attr in inspect.getmembers(plug_type,
+                                            inspect.isdatadescriptor):
+    if isinstance(attr, PlugDependency):
+      yield attr_name, attr.plug_type
+
+
+class _PlugDependencyData(object):
```

Comment:
Hmm, the only changes I'd make before putting it into openhtf proper is:

1) Replace the descriptors concept with a single attribute, maybe `dependencies` or `requires`?
2) Use C3 linearization (which is actually how Python's MRO is calculated): https://en.wikipedia.org/wiki/C3_linearization
3) Obviously have the `PlugManager` handle it

But those aren't necessary for this to get into contrib/

Reply:
(2) This seems to try to get an order to run the initialization in, but here we need to worry about that.  Instead, we just want to get new plugs that are only referenced as plug dependencies.

Unless you mean that each plug should actually be initialized in the specified order.  That could could provide a way to set the attributes before __init__ like the logger attribute.

Reply:
I would add beefing up the docstrings to the list of requests for non-contrib inclusion.